File: backend/utils/recipe_description_cleaner.py
#!/usr/bin/env python3
"""
Recipe Description Cleaner

This utility cleans up Spoonacular recipe descriptions by removing:
- HTML markup
- Pricing information
- Marketing text
- Spoonacular scores and links
- Unnecessary promotional content

It creates clean, simple descriptions focused on the recipe itself.
"""

import re
import html
from typing import Optional

class RecipeDescriptionCleaner:
    """Clean and simplify recipe descriptions"""
    
    def __init__(self):
        # Patterns to remove from descriptions
        self.patterns_to_remove = [
            # HTML tags
            r'<[^>]+>',
            
            # Pricing information
            r'For <b>\$[\d.]+ per serving</b>',
            r'\$[\d.]+ per serving',
            
            # Marketing percentages
            r'<b>covers \d+%</b> of your daily requirements',
            r'covers \d+% of your daily requirements',
            
            # Calorie and macro marketing
            r'Watching your figure\? This .*? recipe has <b>[\d.]+ calories</b>, <b>[\d.]+g of protein</b>, and <b>[\d.]+g of fat</b> per serving\.',
            r'This .*? recipe has <b>[\d.]+ calories</b>, <b>[\d.]+g of protein</b>, and <b>[\d.]+g of fat</b> per serving\.',
            
            # Serving and rating info
            r'This recipe serves \d+\.',
            r'\d+ people have tried and liked this recipe\.',
            
            # Cuisine marketing
            r'This recipe is typical of .*? cuisine\.',
            
            # Spoonacular scores and links
            r'Overall, this recipe earns a <b>solid spoonacular score of \d+%</b>',
            r'spoonacular score of \d+%',
            r'<a href="https://spoonacular\.com/recipes/.*?">.*?</a>',
            r'https://spoonacular\.com/recipes/.*?',
            
            # Similar recipe links
            r'.*? are very similar to this recipe\.',
            
            # Foodista attribution
            r'It is brought to you by Foodista\.',
            
            # Generic marketing phrases
            r'This .*? recipe',
            r'A mixture of .*? ingredients are all it takes to make this recipe so delicious\.',
            r'From preparation to the plate, this recipe takes roughly <b>[\d.]+ minutes</b>\.',
        ]
        
        # Patterns to extract useful information
        self.extract_patterns = {
            'cooking_time': r'<b>(\d+) minutes</b>',
            'calories': r'<b>(\d+) calories</b>',
            'protein': r'<b>(\d+g) of protein</b>',
            'fat': r'<b>(\d+g) of fat</b>',
            'cuisine': r'typical of (\w+) cuisine',
            'ingredients': r'A mixture of (.*?) ingredients',
            'ingredients_alt': r'ingredients are all it takes to make this recipe so delicious',
        }
    
    def clean_description(self, description: str) -> str:
        """
        Clean a recipe description by removing HTML and marketing content.
        
        Args:
            description: Raw description from Spoonacular
            
        Returns:
            Clean, simple description
        """
        if not description:
            return ""
        
        # Decode HTML entities
        cleaned = html.unescape(description)
        
        # Remove all unwanted patterns
        for pattern in self.patterns_to_remove:
            cleaned = re.sub(pattern, '', cleaned, flags=re.IGNORECASE)
        
        # Extract useful information
        useful_info = self._extract_useful_info(description)
        
        # Create a clean description
        clean_desc = self._create_clean_description(cleaned, useful_info)
        
        # Final cleanup
        clean_desc = self._final_cleanup(clean_desc)
        
        # If the cleaned description is too short or doesn't make sense, use fallback
        if len(clean_desc) < 30 or clean_desc.count(' ') < 5:
            clean_desc = self._generate_fallback_description(description, useful_info)
        
        return clean_desc
    
    def _extract_useful_info(self, description: str) -> dict:
        """Extract useful information from the description"""
        info = {}
        
        for key, pattern in self.extract_patterns.items():
            match = re.search(pattern, description, re.IGNORECASE)
            if match:
                # Check if the pattern has capture groups
                if '(' in pattern and ')' in pattern:
                    try:
                        info[key] = match.group(1)
                    except IndexError:
                        # No capture group, skip this pattern
                        continue
                else:
                    # Pattern without capture groups, just mark as found
                    info[key] = True
        
        return info
    
    def _create_clean_description(self, cleaned_text: str, useful_info: dict) -> str:
        """Create a clean description using extracted information"""
        # Build a natural description
        description_parts = []
        
        # Start with recipe type
        if useful_info.get('cuisine'):
            description_parts.append(f"A classic {useful_info['cuisine']} dish")
        else:
            description_parts.append("A delicious homemade dish")
        
        # Add cooking time
        if useful_info.get('cooking_time'):
            description_parts.append(f"that takes approximately {useful_info['cooking_time']} minutes to prepare")
        
        # Add ingredient highlights
        if useful_info.get('ingredients'):
            ingredients = useful_info['ingredients']
            # Clean up ingredients
            ingredients = re.sub(r',+', ',', ingredients)
            ingredients = ingredients.strip(', ')
            if ingredients and len(ingredients) > 5:
                description_parts.append(f"featuring fresh {ingredients}")
        
        # Add nutrition summary
        nutrition_info = []
        if useful_info.get('calories'):
            nutrition_info.append(f"{useful_info['calories']} calories")
        if useful_info.get('protein'):
            nutrition_info.append(f"{useful_info['protein']} protein")
        if useful_info.get('fat'):
            nutrition_info.append(f"{useful_info['fat']}")
        
        if nutrition_info:
            description_parts.append(f"with {', '.join(nutrition_info)} per serving")
        
        # Combine all parts
        if description_parts:
            description = ' '.join(description_parts) + "."
        else:
            description = "A delicious homemade recipe perfect for any meal."
        
        # Ensure proper capitalization
        description = description[0].upper() + description[1:]
        
        return description
    
    def _final_cleanup(self, description: str) -> str:
        """Final cleanup of the description"""
        # Remove extra whitespace
        description = re.sub(r'\s+', ' ', description)
        
        # Remove leading/trailing whitespace
        description = description.strip()
        
        # Remove any remaining HTML-like content
        description = re.sub(r'<[^>]*>', '', description)
        
        # Remove any remaining URLs
        description = re.sub(r'https?://[^\s]+', '', description)
        
        # Clean up punctuation
        description = re.sub(r'\.+', '.', description)
        description = re.sub(r'\s+', ' ', description)
        
        # Ensure it ends with a period
        if description and not description.endswith('.'):
            description += '.'
        
        return description
    
    def _generate_fallback_description(self, original_description: str, useful_info: dict) -> str:
        """Generate a fallback description when cleaning doesn't work well"""
        parts = []
        
        # Try to extract recipe name from the beginning
        recipe_name_match = re.search(r'^([^.!?]+)', original_description)
        if recipe_name_match:
            recipe_name = recipe_name_match.group(1).strip()
            # Clean up the recipe name
            recipe_name = re.sub(r'<[^>]+>', '', recipe_name)
            recipe_name = re.sub(r'\s+', ' ', recipe_name).strip()
            if recipe_name and len(recipe_name) > 10:
                parts.append(f"{recipe_name}")
        
        # Add cuisine if available
        if useful_info.get('cuisine'):
            parts.append(f"is a classic {useful_info['cuisine']} dish")
        else:
            parts.append("is a delicious homemade dish")
        
        # Add cooking time if available
        if useful_info.get('cooking_time'):
            parts.append(f"that takes about {useful_info['cooking_time']} minutes to prepare")
        
        # Add nutrition info if available
        nutrition_parts = []
        if useful_info.get('calories'):
            nutrition_parts.append(f"{useful_info['calories']} calories")
        if useful_info.get('protein'):
            nutrition_parts.append(f"{useful_info['protein']} protein")
        if useful_info.get('fat'):
            nutrition_parts.append(f"{useful_info['fat']}")
        
        if nutrition_parts:
            parts.append(f"with {', '.join(nutrition_parts)} per serving")
        
        # Combine parts
        if parts:
            description = ' '.join(parts) + "."
        else:
            description = "A delicious homemade recipe perfect for any meal."
        
        # Capitalize first letter
        description = description[0].upper() + description[1:]
        
        return description
    
    def clean_multiple_descriptions(self, descriptions: list) -> list:
        """Clean multiple descriptions at once"""
        return [self.clean_description(desc) for desc in descriptions if desc]

# Example usage and testing
def test_description_cleaner():
    """Test the description cleaner with sample data"""
    cleaner = RecipeDescriptionCleaner()
    
    # Sample problematic description
    sample_description = """Easy Eggplant Parmesan might be a good recipe to expand your main course recipe box. For <b>$2.63 per serving</b>, this recipe <b>covers 21%</b> of your daily requirements of vitamins and minerals. Watching your figure? This gluten free, lacto ovo vegetarian, and primal recipe has <b>288 calories</b>, <b>21g of protein</b>, and <b>12g of fat</b> per serving. This recipe serves 4. 2 people have tried and liked this recipe. This recipe is typical of Mediterranean cuisine. A mixture of style cheese, onion, pasta sauce, and a handful of other ingredients are all it takes to make this recipe so delicious. From preparation to the plate, this recipe takes roughly <b>45 minutes</b>. It is brought to you by Foodista. Overall, this recipe earns a <b>solid spoonacular score of 64%</b>. <a href="https://spoonacular.com/recipes/easy-eggplant-parmesan-265330">Easy Eggplant Parmesan</a>, <a href="https://spoonacular.com/recipes/easy-eggplant-parmesan-1620099">Easy Eggplant Parmesan</a>, and <a href="https://spoonacular.com/recipes/easy-eggplant-parmesan-103890">Easy Eggplant Parmesan</a> are very similar to this recipe."""
    
    print("🧪 Testing Recipe Description Cleaner")
    print("=" * 50)
    
    print("\n📝 Original Description:")
    print(sample_description)
    
    print("\n✨ Cleaned Description:")
    cleaned = cleaner.clean_description(sample_description)
    print(cleaned)
    
    print("\n📊 Length Comparison:")
    print(f"Original: {len(sample_description)} characters")
    print(f"Cleaned: {len(cleaned)} characters")
    print(f"Reduction: {((len(sample_description) - len(cleaned)) / len(sample_description) * 100):.1f}%")

if __name__ == "__main__":
    test_description_cleaner()

File: src/utils/recipeDescriptionCleaner.ts
/**
 * Recipe Description Cleaner
 * 
 * This utility cleans up Spoonacular recipe descriptions by removing:
 * - HTML markup
 * - Pricing information
 * - Marketing text
 * - Spoonacular scores and links
 * - Unnecessary promotional content
 * 
 * It creates clean, simple descriptions focused on the recipe itself.
 */

export class RecipeDescriptionCleaner {
  private patternsToRemove: RegExp[] = [
    // HTML tags
    /<[^>]+>/g,
    
    // Pricing information
    /For <b>\$[\d.]+ per serving<\/b>/gi,
    /\$[\d.]+ per serving/gi,
    
    // Marketing percentages
    /<b>covers \d+%<\/b> of your daily requirements/gi,
    /covers \d+% of your daily requirements/gi,
    
    // Calorie and macro marketing
    /Watching your figure\? This .*? recipe has <b>[\d.]+ calories<\/b>, <b>[\d.]+g of protein<\/b>, and <b>[\d.]+g of fat<\/b> per serving\./gi,
    /This .*? recipe has <b>[\d.]+ calories<\/b>, <b>[\d.]+g of protein<\/b>, and <b>[\d.]+g of fat<\/b> per serving\./gi,
    
    // Serving and rating info
    /This recipe serves \d+\./gi,
    /\d+ people have tried and liked this recipe\./gi,
    
    // Cuisine marketing
    /This recipe is typical of .*? cuisine\./gi,
    
    // Spoonacular scores and links
    /Overall, this recipe earns a <b>solid spoonacular score of \d+%<\/b>/gi,
    /spoonacular score of \d+%/gi,
    /<a href="https:\/\/spoonacular\.com\/recipes\/.*?">.*?<\/a>/gi,
    /https:\/\/spoonacular\.com\/recipes\/.*?/gi,
    
    // Similar recipe links
    /.*? are very similar to this recipe\./gi,
    
    // Foodista attribution
    /It is brought to you by Foodista\./gi,
    
    // Generic marketing phrases
    /This .*? recipe/gi,
    /A mixture of .*? ingredients are all it takes to make this recipe so delicious\./gi,
    /From preparation to the plate, this recipe takes roughly <b>[\d.]+ minutes<\/b>\./gi,
  ];

  private extractPatterns: { [key: string]: RegExp } = {
    cookingTime: /<b>(\d+) minutes<\/b>/gi,
    cuisine: /typical of (\w+) cuisine/gi,
    ingredients: /A mixture of (.*?) ingredients/gi,
  };

  /**
   * Clean a recipe description by removing HTML and marketing content.
   */
  cleanDescription(description: string): string {
    if (!description) return "";

    // Decode HTML entities
    let cleaned = this.decodeHtmlEntities(description);

    // Remove only the most problematic patterns, keep more content
    const aggressivePatterns = [
      // HTML tags
      /<[^>]+>/g,
      
      // Most annoying marketing text
      /Watching your figure\?/gi,
      /For <b>\$[\d.]+ per serving<\/b>/gi,
      /covers \d+% of your daily requirements/gi,
      /It is brought to you by Foodista\./gi,
      /With a spoonacular score of \d+%/gi,
      /Users who liked this recipe also liked.*?\./gi,
      /This recipe is liked by \d+ foodies and cooks\./gi,
      /This dish is solid\./gi,
    ];

    // Apply aggressive cleaning
    for (const pattern of aggressivePatterns) {
      cleaned = cleaned.replace(pattern, '');
    }

    // Extract useful information (just cuisine and cooking time, no macros)
    const usefulInfo = this.extractUsefulInfo(description);

    // Create a clean description
    let cleanDesc = this.createCleanDescription(cleaned, usefulInfo);

    // Final cleanup
    cleanDesc = this.finalCleanup(cleanDesc);

    // If the cleaned description is too short, use fallback
    if (cleanDesc.length < 50 || cleanDesc.split(' ').length < 8) {
      cleanDesc = this.generateFallbackDescription(description, usefulInfo);
    }

    return cleanDesc;
  }

  /**
   * Clean multiple descriptions at once
   */
  cleanMultipleDescriptions(descriptions: string[]): string[] {
    return descriptions.map(desc => this.cleanDescription(desc)).filter(desc => desc);
  }

  private decodeHtmlEntities(text: string): string {
    const textarea = document.createElement('textarea');
    textarea.innerHTML = text;
    return textarea.value;
  }

  private extractUsefulInfo(description: string): { [key: string]: string } {
    const info: { [key: string]: string } = {};

    for (const [key, pattern] of Object.entries(this.extractPatterns)) {
      const match = pattern.exec(description);
      if (match && match[1]) {
        info[key] = match[1];
      }
    }

    return info;
  }

  private createCleanDescription(cleanedText: string, usefulInfo: { [key: string]: string }): string {
    // Instead of creating a very short description, let's clean the original text
    // and make it more natural while keeping more content
    
    let description = cleanedText;
    
    // Clean up common issues
    description = description.replace(/\s+/g, ' '); // Fix multiple spaces
    description = description.replace(/\.+/g, '.'); // Fix multiple periods
    description = description.replace(/,\s*,/g, ','); // Fix double commas
    
    // Remove any remaining marketing phrases
    description = description.replace(/Head to the store and pick up.*?today\./gi, '');
    description = description.replace(/This recipe is liked by \d+ foodies and cooks\./gi, '');
    description = description.replace(/This dish is solid\./gi, '');
    
    // Clean up the beginning to make it more natural
    description = description.replace(/^([^.!?]+) - ([^.!?]+) requires about/gi, '$1 ($2) takes about');
    description = description.replace(/^([^.!?]+) requires about/gi, '$1 takes about');
    
    // Make it more natural
    description = description.replace(/This lacto ovo vegetarian recipe has/gi, 'This vegetarian recipe features');
    description = description.replace(/This recipe has/gi, 'This recipe features');
    
    // Clean up the end
    description = description.replace(/\.\s*$/, ''); // Remove trailing period if exists
    
    // Ensure it ends with a period
    if (description && !description.endsWith('.')) {
      description += '.';
    }
    
    // If the description is still too short, add some context
    if (description.length < 100) {
      if (usefulInfo.cuisine) {
        description = `A classic ${usefulInfo.cuisine} dish. ${description}`;
      } else {
        description = `A delicious homemade recipe. ${description}`;
      }
    }
    
    return description;
  }

  private generateFallbackDescription(originalDescription: string, usefulInfo: { [key: string]: string }): string {
    // Create a more natural description from the original text
    let description = originalDescription;
    
    // Remove HTML tags but keep the text content
    description = description.replace(/<[^>]+>/g, '');
    
    // Remove the most problematic marketing text
    description = description.replace(/Watching your figure\?/gi, '');
    description = description.replace(/For \$\d+\.\d+ per serving/gi, '');
    description = description.replace(/\$\d+\.\d+ per serving/gi, '');
    description = description.replace(/covers \d+% of your daily requirements/gi, '');
    description = description.replace(/It is brought to you by Foodista\./gi, '');
    description = description.replace(/With a spoonacular score of \d+%/gi, '');
    description = description.replace(/Users who liked this recipe also liked.*?\./gi, '');
    description = description.replace(/This recipe is liked by \d+ foodies and cooks\./gi, '');
    description = description.replace(/This dish is solid\./gi, '');
    
    // Clean up the text
    description = description.replace(/\s+/g, ' ').trim();
    description = description.replace(/\.+/g, '.');
    
    // Make it more natural
    description = description.replace(/This recipe serves \d+\./gi, '');
    description = description.replace(/This recipe is typical of (.*?) cuisine\./gi, 'This is a classic $1 dish.');
    
    // Ensure it ends with a period
    if (description && !description.endsWith('.')) {
      description += '.';
    }
    
    // If the description is still too short, add some context
    if (description.length < 80) {
      if (usefulInfo.cuisine) {
        description = `A classic ${usefulInfo.cuisine} dish. ${description}`;
      } else {
        description = `A delicious homemade recipe. ${description}`;
      }
    }
    
    return description;
  }

  private finalCleanup(description: string): string {
    // Remove extra whitespace
    description = description.replace(/\s+/g, ' ');

    // Remove leading/trailing whitespace
    description = description.trim();

    // Remove any remaining HTML-like content
    description = description.replace(/<[^>]*>/g, '');

    // Remove any remaining URLs
    description = description.replace(/https?:\/\/[^\s]+/g, '');

    // Clean up punctuation
    description = description.replace(/\.+/g, '.');
    description = description.replace(/\s+/g, ' ');

    // Ensure it ends with a period
    if (description && !description.endsWith('.')) {
      description += '.';
    }

    return description;
  }
}

// Create a singleton instance
export const recipeDescriptionCleaner = new RecipeDescriptionCleaner();

// Convenience function
export const cleanRecipeDescription = (description: string): string => {
  return recipeDescriptionCleaner.cleanDescription(description);
};
